Deterministically dispose instances created by WebHostBuilder

pakrym opened the pull request:
Fixes: https://github.com/aspnet/Hosting/issues/867
IServiceProvider would stop disposing instances: https://github.com/aspnet/DependencyInjection/pull/462

@davidfowl @Tratcher

davidfowl commented on src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs:
```diff
@@ -195,9 +195,8 @@ private IServiceCollection BuildHostingServices()
-            var diagnosticSource = new DiagnosticListener("Microsoft.AspNetCore");
-            services.AddSingleton<DiagnosticSource>(diagnosticSource);
-            services.AddSingleton<DiagnosticListener>(diagnosticSource);
+            services.AddSingleton<DiagnosticListener>(provider => new DiagnosticListener("Microsoft.AspNetCore"));
```
Does this need to be disposed?

pakrym replied:
It is IDisposible

pakrym replied:
then we would need two different `BuildHostingServices` or at least parts that need to chain.

pakrym replied:
Also we don't mutate `DiagnosticSource` so it might be fine to return different instances.

davidfowl commented on src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs:
```diff
@@ -257,5 +256,10 @@ private string ResolveContentRootPath(string contentRootPath, string basePath)
+
+        public void Dispose()
+        {
+            _loggerFactory?.Dispose();
```
Shouldn't this be disposed with the host? Why is the builder disposable?

pakrym replied:
No, Builder creates it and will reuse for any number of WebHosts you build using it.

davidfowl commented on src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs:
```diff
@@ -195,9 +195,8 @@ private IServiceCollection BuildHostingServices()
-            services.AddSingleton<DiagnosticSource>(diagnosticSource);
-            services.AddSingleton<DiagnosticListener>(diagnosticSource);
+            services.AddSingleton<DiagnosticListener>(provider => new DiagnosticListener("Microsoft.AspNetCore"));
+            services.AddSingleton<DiagnosticSource>(provider => provider.GetService<DiagnosticListener>());
```
Not sure I like this.

pakrym replied:
We don't have multi interface registration and there is no need to have this as an instance as far as I understand.

Tratcher commented on src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs:
```diff
@@ -22,7 +22,7 @@ namespace Microsoft.AspNetCore.Hosting
     /// A builder for <see cref="IWebHost"/>
     /// </summary>
-    public class WebHostBuilder : IWebHostBuilder
+    public class WebHostBuilder : IWebHostBuilder, IDisposable
```
Should IWebHostBuilder be IDisposable?

pakrym commented:
1. Changed WebHostBuilder to allow creating only single instance of WebHost.
2. LoggerFactory is now disposed with WebHost is it was automatically created.
3. If LoggerFactory was passed in it's not disposed.
4. Added tests for all these cases.

pakrym commented:
Updated again, now there is two methods to add host services and to forward them to application services. Also fixed server disposal issue.

davidfowl commented on src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs:
```diff
@@ -245,6 +262,17 @@ private IServiceCollection BuildHostingServices()
+            services.Replace(ServiceDescriptor.Singleton(typeof(DiagnosticListener), listener));
+            services.Replace(ServiceDescriptor.Singleton(typeof(DiagnosticSource), listener));
+        }
+
```
nit: extra space

davidfowl commented on src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs:
```diff
@@ -245,6 +262,17 @@ private IServiceCollection BuildHostingServices()
 
+        private void AddApplicationServices(IServiceCollection services, IServiceProvider hostingServiceProvider)
+        {
+            var loggerFactory = hostingServiceProvider.GetService<ILoggerFactory>();
```
Add comments here explaining why thungs are so bizzare

davidfowl commented on src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs:
```diff
@@ -175,30 +189,33 @@ private IServiceCollection BuildHostingServices()
             var services = new ServiceCollection();
             services.AddSingleton(_hostingEnvironment);
 
+            //The configured ILoggerFactory is added as a singleton here. AddLogging below will not add an additional one.
```
nit: space after //

pakrym commented:
Also undid change to `UseServer` and fixed tests for it.

davidfowl commented on test/Microsoft.AspNetCore.Hosting.Tests/WebHostBuilderTests.cs:
```diff
@@ -188,6 +189,7 @@ public void MultipleConfigureLoggingInvokedInOrder()
                 .UseStartup<StartupNoServices>();
 
             var host = (WebHost)hostBuilder.Build();
+            var factory = host.Services.GetService<ILoggerFactory>();
```
What is this for?

davidfowl commented on test/Microsoft.AspNetCore.Hosting.Tests/WebHostBuilderTests.cs:
```diff
@@ -494,6 +496,86 @@ public void Configure_SupportsStaticMethodDelegate()
+            var factoryFromHost = host.Services.GetService<ILoggerFactory>();
+            Assert.Equal(factory, factoryFromHost);
+        }
+
```
nit: extra spaces

pakrym replied:
Once `ILoggerFactory` was registered as factory so resolve was required to invoke callbacks. Not true anymore. Reverting.

muratg commented on src/Microsoft.AspNetCore.Hosting/Internal/ServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,20 @@
+
+namespace Microsoft.AspNetCore.Hosting.Internal
+{
+    public static class ServiceCollectionExtensions
```
Actually making the class internal?
